### PR TITLE
Add expiration time to update api key api

### PR DESCRIPTION
**Pull request by jfreden**
This PR adds an optional `expiration` time field to the following APIs: 
- [PUT /_security/api_key/<id>](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-update-api-key.html)
- [PUT /_security/cross_cluster/api_key/<id>](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-update-cross-cluster-api-key.html)
- [POST /_security/api_key/_bulk_update](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-bulk-update-api-keys.html)

Example Request: 
```javascript
PUT /_security/api_key/nlZ6fIwBqia8cKnQkNal
{
    "expiration": "2d"
}
```
Response
```javascript
{
    "updated": true
}
```

I opted to not do datemath since it is not in the create apikey api and would also add significant scope to this PR. If needed it can be added in a separate PR.

### CODE REVIEW

**Comment by jfreden** [comment]
@elasticsearchmachine generate changelog

**Comment by jfreden** [comment]
@elasticsmachine generate changelog

**Comment by jfreden** [comment]
@elasticmachine generate changelog

**Comment by jfreden** [comment]
@elasticsearchmachine generate changelog

File: docs/reference/rest-api/security/bulk-update-api-keys.asciidoc
Comment (n1v0lg) [inline docs/reference/rest-api/security/bulk-update-api-keys.asciidoc:67]: Higher up in the docs, there's a sentence "This API supports updates to API key access scope and metadata" -- lets tweak that to reflect the change. Same for the other API docs.

Could also add that you can omit `expiration` to leave one previously assigned unchanged.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java:65]: I think we need to pick a value other than `instance.getExpiration()`, otherwise we end up potentially not mutating the instance -- `randomValueOtherThan()` should do the trick

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java:26]: let's use `UPDATE_API_KEY_EXPIRATION_TIME_ADDED` -- avoids needing to update the version in two places in case main gets a new version before this is merged

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/BulkUpdateApiKeyRequestSerializationTests.java
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/BulkUpdateApiKeyRequestSerializationTests.java:23]: Nice, thanks for adding this!

**Review by n1v0lg — APPROVED** [review]
LGTM! 

A couple of test tweaks and nits, but nothing that requires another review round.

There's a couple of things to consider for follow ups:

1. We should add `expiration` to the audit log event [here](https://github.com/elastic/elasticsearch/blob/5d2bd02459692a592aae9eb1a9ae1fab32099ee0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java#L1327)
2. Your PR prompted me to think about this again: it's kinda bogus that in the create API you can set 0 or -1 for `expiration` -- we've talked in the past about preventing this but didn't get around to it. I like being consistent between APIs, but I don't think we want to inherit API flaws. WDYT about preventing non-positive values for expiration for the update APIs? 
3. Not sure about this one, but with `metadata` and `role_descriptors` it's possible to "nullify them", i.e., it's possible to set metadata to `{}`. I think it could be nice to give end-users the option to remove expiration by passing `expiration: null` in the request. This would be at odds with generally stirring people away from never-expiring credentials, so I'm not positive. Still, wanted to bring up for discussion so we're making an active choice either way.

**Comment by jfreden** [comment]
Thanks for the review! I'll probably wait with merging this until after the break.

> We should add expiration to the audit log event [here](https://github.com/elastic/elasticsearch/blob/5d2bd02459692a592aae9eb1a9ae1fab32099ee0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java#L1327)

Yes! Will do a follow up with this change.

> Your PR prompted me to think about this again: it's kinda bogus that in the create API you can set 0 or -1 for expiration -- we've talked in the past about preventing this but didn't get around to it. I like being consistent between APIs, but I don't think we want to inherit API flaws. WDYT about preventing non-positive values for expiration for the update APIs?

I agree, I'll add validation in a separate PR for all update/create api key APIs. 

> Not sure about this one, but with metadata and role_descriptors it's possible to "nullify them", i.e., it's possible to set metadata to {}. I think it could be nice to give end-users the option to remove expiration by passing expiration: null in the request. This would be at odds with generally stirring people away from never-expiring credentials, so I'm not positive. Still, wanted to bring up for discussion so we're making an active choice either way.

Yes, hmm. I'm on the fence but I think I'm leaning towards not allowing it, since generally stirring people away from never-expiring credentials feels like the correct approach, like you said.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java
Comment (jfreden) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java:26]: I picked this one since it needs to be an old version when we're testing backwards compatibility (`UPDATE_API_KEY_EXPIRATION_TIME_ADDED` is the current version).

Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestSerializationTests.java:26]: ahh right. makes sense!

**Comment by n1v0lg** [comment]
@jfreden 

> I agree, I'll add validation in a separate PR for all update/create api key APIs.

perfect, thanks! let's start with just update -- the create API is an existing API and while I don't think it's a breaking change, it's breaking change adjacent so we'd at least have to take a closer look, whereas with the update API, the field is new so we can add w/e validation we want there

**Comment by jfreden** [comment]
We can explore it more in the follow up PR but if feels like not allowing creation of expired api keys would make sense/be pretty safe.

**Pull request merged**
